aptos-move/move-examples/token_objects/sources/collection.move

```diff
@@ -0,0 +1,439 @@
+/// This defines an object-based Collection. A collection acts as a set organizer for a group of
+/// tokens. This includes aspects such as a general description, project URI, name, and may contain
+/// other useful generalizations across this set of tokens.
+///
+/// Being built upon objects enables collections to be relatively flexible. As core primitives it
+/// supports:
+/// * Common fields: name, uri, description, creator
+/// * A mutability config for uri and description
+/// * Optional support for collection-wide royalties
+/// * Optional support for tracking of supply
+///
+/// This collection does not directly support:
+/// * Events on mint or burn -- that's left to the collection creator.
+///
+/// TODO:
+/// * Add Royalty reading and consider mutation
+/// * Consider supporting changing the name of the collection.
+/// * Consider supporting changing the aspects of supply
+/// * Add aggregator support when added to framework
+/// * Update ObjectId to be an acceptable param to move
+module token_objects::collection {
+    use std::error;
+    use std::option::{Self, Option};
+    use std::signer;
+    use std::string::{Self, String};
+
+    use aptos_framework::object::{Self, CreatorRef, ObjectId};
+
+    friend token_objects::token;
+
+    /// The collections supply is at its maximum amount
+    const EEXCEEDS_MAX_SUPPLY: u64 = 0;
+    /// The collection does not exist
+    const ECOLLECTION_DOES_NOT_EXIST: u64 = 1;
+    /// The provided signer is not the creator
+    const ENOT_CREATOR: u64 = 2;
+    /// Attempted to mutate an immutable field
+    const EFIELD_NOT_MUTABLE: u64 = 3;
+
+    #[resource_group_member(group = aptos_framework::object::ObjectGroup)]
+    /// Represents the common fields for a collection.
+    struct Collection has key {
+        /// The creator of this collection.
+        creator: address,
+        /// A brief description of the collection.
+        description: String,
+        /// Determines which fields are mutable.
+        mutability_config: MutabilityConfig,
+        /// An optional categorization of similar token.
+        name: String,
+        /// The Uniform Resource Identifier (uri) pointing to the JSON file stored in off-chain
+        /// storage; the URL length will likely need a maximum any suggestions?
+        uri: String,
+    }
+
+    /// This config specifies which fields in the TokenData are mutable
+    struct MutabilityConfig has copy, drop, store {
+        description: bool,
+        uri: bool,
+    }
+
+    #[resource_group_member(group = aptos_framework::object::ObjectGroup)]
+    /// The royalty of a token within this collection -- this optional
+    struct Royalty has drop, key {
+        numerator: u64,
```

Comment:
Any particular reason why we not use FixedPoint32? Do we need to introduce FixedPoint64? 

If there is logic in the program which needs to reason about this an abstraction is highly desirable for keeping the Prover happy.

Reply:
Hmm... I guess we could, let me look at the API for it :)